From 694ac3b785aa11efc4134ee4920fdee1e7872df0 Mon Sep 17 00:00:00 2001
From: Alex Belits <alex.belits@cavium.com>
Date: Thu, 18 Oct 2018 10:45:57 +0300
Subject: [PATCH 0315/1051] octeontx: Add task cleanup callback in GPIO
 interrupt functionality.

Signed-off-by: Alex Belits <Alex.Belits@cavium.com>
Signed-off-by: Yury Norov <ynorov@marvell.com>
Signed-off-by: Yury Norov <ynorov@caviumnetworks.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../cavium/octeontx-83xx/octeontx_main.c       | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c b/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
index 4babd7d2f7a6..29d946d25cec 100644
--- a/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
+++ b/drivers/net/ethernet/cavium/octeontx-83xx/octeontx_main.c
@@ -1217,15 +1217,8 @@ static int octtx_dev_open(struct inode *inode, struct file *fp)
 
 static int octtx_dev_release(struct inode *inode, struct file *fp)
 {
-	int i;
-
 	if (gpio_in_use == 0)
 		return -EINVAL;
-
-	for (i = 0; i < MAX_GPIO; i++)
-		if (gpio_installed[i] != 0)
-			__remove_el3_inthandler(i);
-
 	gpio_in_use = 0;
 	return 0;
 }
@@ -1308,6 +1301,13 @@ static int __init octeontx_init_module(void)
 	queue_delayed_work(check_link, &dwork, 0);
 	queue_delayed_work(reset_domain, &dwork_reset, 0);
 
+	/* Register task cleanup handler */
+	ret = task_cleanup_handler_add(cleanup_el3_irqs);
+	if (ret != 0) {
+		ret = -ENODEV;
+		goto cleanup_handler_err;
+	}
+
 	/* create a char device */
 	ret = alloc_chrdev_region(&octtx_dev, 1, 1, DEVICE_NAME);
 	if (ret != 0) {
@@ -1366,6 +1366,9 @@ static int __init octeontx_init_module(void)
 	unregister_chrdev_region(octtx_dev, 1);
 
 alloc_chrdev_err:
+cleanup_handler_err:
+	task_cleanup_handler_remove(cleanup_el3_irqs);
+
 wq_err:
 	symbol_put(timpf_com);
 
@@ -1418,6 +1421,7 @@ static void __exit octeontx_cleanup_module(void)
 	symbol_put(timpf_com);
 	symbol_put(lbk_com);
 	symbol_put(thunder_bgx_com);
+	task_cleanup_handler_remove(cleanup_el3_irqs);
 }
 
 module_init(octeontx_init_module);
-- 
2.17.1

